### PR TITLE
[Dotnet] port of ScanningAndDetecting3DObjects

### PR DESCRIPTION
**Pull request by stephen-hawley**
This is a huge sample, thanks for your patience.

### CODE REVIEW

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+	bool disposed = false;
+	override protected void Dispose (bool disposing)
+	{
+		if (!disposed)
+		{
+			disposed = true;
+			if (disposing)
+			{
+				base.Dispose ();
```
Comment (rolfbjarne) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs:346]: The Dispose pattern is wrong, the `Dispose (bool)` method should call the same base method, so:

```
base.Dispose (disposing);
```

This effectively changes the whole method, and the comment in the README about Xamarin's weird Dispose pattern is also wrong (the original author got the Dispose pattern wrong).

This method would become:

```csharp
if (!disposed)
{
    disposed = true;
    NSNotificationCenter.DefaultCenter.RemoveObserver (notificationObserverHandle);
    Origin?.Dispose ();
    BoundingBox?.Dispose ();
    GhostBoundingBox?.Dispose ();
}
base.Dispose (disposing);
```

Ref: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implement-the-dispose-pattern-for-a-derived-class

This occurs in other files as well.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	{
+		this.busyCreatingReferenceObject = isBusyCreatingReferenceObject;
+	}
+
+	internal new void Dispose ()
+	{
+		Dispose (true);
+		GC.SuppressFinalize (this);
+	}
+
```
[the remark this answers is not in the document]
Comment (stephen-hawley) [reply dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs:337]: I think I'm just going to remove these. They were sketchy anyway. At least one was marked as being put in for debugging.

**Review by rolfbjarne — APPROVED** [review]

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+		NSNotificationCenter.DefaultCenter.PostNotificationName (ScannedObject.BoundingBoxCreatedNotificationName, null);
+	}
+
+	void UpdateOrCreateGhostBoundingBox ()
+	{
+		// Perform a hit test against the feature point cloud. 
+		var result = sceneView.SmartHitTest (ViewController.Instance.ScreenCenter);
+		if (result is null)
+		{
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs:174]: Not sure if worth going through many of these, but mono guidelines recommend `if () {` as well as the else statement

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+		var bbwp = boundingBox.WorldPosition;
+		boundingBox.WorldPosition = new SCNVector3 (bbwp.X, bbwp.Y - diffOnY / 2, bbwp.Z);
+	}
+
+
+
+
+	void ScanningStateChanged (NSNotification notification)
+	{
+		var state = notification.UserInfo [Scan.StateUserInfoKey] as SimpleBox<Scan.ScanState>;
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedObject.cs:291]: good place for is operator

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedPointCloud.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		var boundingBox = scannedObject.BoundingBox ?? scannedObject.GhostBoundingBox;
+		if (boundingBox is not null && !boundingBox.IsDisposed)
+		{
+			UpdateBoundingBox (boundingBox);
+		}
+	}
+
+	void ScanningStateChanged (NSNotification notification)
+	{
+		var state = notification.UserInfo? [Scan.StateUserInfoKey] as SimpleBox<Scan.ScanState>;
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/ScannedPointCloud.cs:63]: is operator would be good here

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/Tile.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+		}
+
+		internal void UpdateVisualization()
+		{
+			var newOpacity = Captured ? 0.5 : 0.0;
+			newOpacity += Highlighted ? 0.35 : 0.0;
+			Opacity = (nfloat) newOpacity;
+		}
+	}
+}
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/3D_UI/Tile.cs:34]: new line

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/Controllers/TestRun.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	protected override void Dispose (bool disposing)
+	{
+		if (!disposed)
+		{
+			disposed = true;
+			if (disposing)
+			{
+				DetectedObject?.RemoveFromParentNode ();
+
+				if (sceneView.Session.Configuration as ARWorldTrackingConfiguration is not null)
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/Controllers/TestRun.cs:135]:
```suggestion
				if (sceneView.Session.Configuration as ARWorldTrackingConfiguration)
```
Think this is redundant

File: dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/Controllers/ViewControllerNavigationBar.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+	internal ViewControllerNavigationBar (UINavigationBar navigationBar, EventHandler previousButtonTapped, EventHandler restartButtonTapped)
+	{
+		this.navigationBar = navigationBar;
+
+		backButton = new UIBarButtonItem ("Back", UIBarButtonItemStyle.Plain, previousButtonTapped);
+		startOverButton = new UIBarButtonItem ("Restart", UIBarButtonItemStyle.Plain, restartButtonTapped);
+
+		var navigationItem = new UINavigationItem ("Start");
+		navigationItem.LeftBarButtonItem = backButton;
```
Comment (tj-devel709) [inline dotnet/ios/ios12/ScanningAndDetecting3DObjects/ScanningApp/Controllers/ViewControllerNavigationBar.cs:17]: Can these be added to the constructor?

**Comment by rolfbjarne** [comment]
[Microsoft support for Xamarin](https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin) will end on May 1, 2024 for all Xamarin SDKs. In preparation for this, all issues and PRs in this repository are being closed.

**Pull request closed without merging**